sklearn/utils/tests/test_estimator_checks.py

```diff
@@ -628,6 +628,13 @@ class NonConformantEstimatorNoParamSet(BaseEstimator):
         def __init__(self, you_should_set_this_=None):
             pass
 
+    class ConformantEstimatorWithExtraParams(BaseEstimator):
+        def _more_tags(self):
+            return {"extra_params": ["param1"]}
```

Comment:
A first thought: this is hard to extend with inheritance.

Reply:
should we go through MRO and handle extending attributes? I think @amueller wasn't very happy about a manual MRO walk. Also, WDYT @rth ?

Reply:
> I think @amueller wasn't very happy about a manual MRO walk

I am not sure that you can otherwise if one wants to inherit `extra_params` here.
The problem spotted here would also be true for other tags that would accept an iterable, e.g. `preserve_dtypes` and `X_types`.

Reply:
In any case, that's not an issue which we'd like to solve in this PR I think.

Reply:
Yes, I agree it's an issue for any tag with a collection for a value, but I'm not sure that those other examples would warrant the use of inheritance to extend them as much as the present case, since "what gets cloned" is likely to be the sort of thing added to by a mixin. One solution is to use a prefixing system with scalar values rather than a collection value. But I'm okay leaving this concern alone, too.

Reply:
> should we go through MRO and handle extending attributes? I think amueller wasn't very happy about a manual MRO walk.

Well it's more that if we can keep it simple and avoid manual MRO walk we could as well do that. 

Right now it's not clear to me if this would be used anywhere else than for '_metadata_request' in practice, to evaluate whether inheritance, or lack of it would be an issue.  +1 to keep the current version for now.

Reply:
> Right now it's not clear to me if this would be used anywhere else than for '_metadata_request' in practice, 

Yep, at least inside sklearn, I don't see us adding anything else anytime soon, or too frequently at all. And however it's used by external libraries, I'm sure they can find a workaround to the issue.

Reply:
> And however it's used by external libraries, I'm sure they can find a workaround to the issue.

I think we need to at least provide some guidance and what this work around looks like.

We can special case `non_init_params` here:

https://github.com/scikit-learn/scikit-learn/blob/4b0d291fb30d057980a9bf64331c398f7c425fed/sklearn/base.py#L348-L349

and have it actual extend the list.

Are there any use cases we see for `non_init_params` besides `_metadata_request`?

Reply:
> We can special case non_init_params here:

It'll make that part of the code a bit dirty, but I won't oppose. Should it be included in this PR?

> I think we need to at least provide some guidance and what this work around looks like.

Since the workaround would apply to all lists, should we have it in a different PR?

> Are there any use cases we see for non_init_params besides _metadata_request?

As far as I'm concerned, `metadata_request` is the only one, but I guess third party developers can get creative.

Reply:
> It'll make that part of the code a bit dirty, but I won't oppose. Should it be included in this PR?

I am thinking if that is the workaround third parties would have to do.

> I guess third party developers can get creative.

I am starting to see it as a hack to get freezing to work. Imagine listing out all the attributes you want to "freeze" in `non_init_params`.

> As far as I'm concerned, metadata_request is the only one

In #20350, was there any reason for a developer to know that `_metadata_request` exist and need to interact with `metadata_request` directly?

Reply:
> I am thinking if that is the workaround third parties would have to do.

The workaround would be for them to get the tags in `_more_tags`, and make sure the extend it, rather than returning only new ones, which is not a bad hack.

> I am starting to see it as a hack to get freezing to work. Imagine listing out all the attributes you want to "freeze" in non_init_params.

It'll probably involve more than that for meta-estimators, but I like the idea.

> In #20350, was there any reason for a developer to know that _metadata_request exist and need to interact with metadata_request directly?

I don't think so, they pretty much would work with the helper functions and classes.